Домашнее задание 11 на проверку

PavelRomashin opened the pull request:
Работа на проверку

PavelRomashin committed:
Домашнее задание 11 на проверку

kirillNay commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -6,17 +6,39 @@ import android.os.Bundle
+    private val ITunesApiBaseUrl = "https://itunes.apple.com"
+    private val retrofit = Retrofit.Builder().baseUrl(ITunesApiBaseUrl)
+        .addConverterFactory(GsonConverterFactory.create()).build()
+    private val notFound = "Ничего не нашлось"
```
⏫ Стоит всегда помнить о необходимости выносить строки в строковые ресуры - `strings.xml`, как это было в спринте 9

kirillNay commented on app/src/main/java/com/example/playlistmaker/TrackHolder.kt:
```diff
@@ -16,7 +18,7 @@ class TrackHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
         trackName.text = model.trackName
         artistName.text = model.artistName
-        trackLength.text = model.trackTime
+        trackLength.text = SimpleDateFormat("mm:ss", Locale.getDefault()).format(293000L)
```
⚠️ Кажется, в метод `format` передается значение из примера. Стоит заменить его на значение из передаваемой модели трека, иначе все треки в плейлисте будут продолжительностью в 04:53))

kirillNay commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -64,6 +100,59 @@ class SearchActivity : AppCompatActivity() {
+            trackAdapter.notifyDataSetChanged()
+            recyclerView.visibility = View.GONE
+
+            if (text == notFound) {
```
⏫ Чтобы избежать сравнение `String`,  разработчики часто используют следующий подход:
Создают общий класс ошибки, обычно он является `sealed` классом (его еще можно назвать UI State ошибки), в котором указываются способы отображения той или иной ошибки (текст ошибки, изображение ошибки, видна ли кнопка "Обновить" и т.п.).
Затем  создают под каждый вид ошибки отдельный класс или `object`, который наследуется от общего класса. 

Вот пример того, как такой класс может выглядеть.
```

sealed class UIError(
            @StringRes val text: Int,
            @DrawableRes val icon: Int,
            isButtonVisible: Boolean
    ) {
        
        object NotFound : UIError(...)

        ...
    }
```

Метод `showMessage` может лишь принимать объект класса `UIError` и в зависимости от его параметров менять состояния вьюшек

kirillNay commented on app/src/main/java/com/example/playlistmaker/Track.kt:
```diff
@@ -1,37 +1,12 @@
-        "https://is5-ssl.mzstatic.com/image/thumb/Music125/v4/a0/4d/c4/a04dc484-03cc-02aa-fa82-5334fcb4bc16/18UMGIM24878.rgb.jpg/100x100bb.jpg"
-    )
-)
+val trackLibrary = ArrayList<Track>()
```
⏫ В данном случае список треков хранится в глобальной переменной и будет жить всю жизнь приложения. Даже если мы перезайдем на экран "Поиска", мы все равно увидим старый запрос, хотя в поисковой строке пусто.
Советую перенести хранение `trackLibrary` из глобальной переменной в переменную класса `SearchActivity`. 

Но не стоит забывать о необходимости сохранять значение списка при перевороте экрана!

kirillNay commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -64,6 +100,59 @@ class SearchActivity : AppCompatActivity() {
+        }
+    }
+
+    private fun showMessage(text: String) {
```
⏫ После переворота экрана у нас пропадает сообщение об ошибке до тех пор, пока снова не нажмем на кнопку DONE на клавиатуре. Стоит подумать над сохранением состояния экрана при смене конфигурации

kirillNay commented on app/src/main/res/layout-v28/activity_search.xml:
```diff
@@ -60,8 +62,56 @@
+                    android:layout_height="wrap_content"
+                    android:layout_gravity="center"
+                    android:layout_marginTop="@dimen/double_track_layout_margin"
+                    android:fontFamily="@font/ys_display_medium"
```
⏫ Скорее всего в приложении на разных экранах будут отображаться экраны с ошибками в одном стиле. Чтобы поддержать общую стилистику приложения можем вынести атрибуты оформления в отдельный стиль

kirillNay commented on app/src/main/res/layout-v28/activity_search.xml:
```diff
@@ -18,7 +18,7 @@
         android:layout_height="fill_parent"
         android:orientation="vertical"
-        android:paddingHorizontal="8dp"
+        android:paddingHorizontal="16dp"
```
⏫ `layout-v28` используется, если мы хотим поддержать отдельную верстку для устройств с версией SDK-28 и больше. Такое может быть полезно, если например мы используем элементы с атрибутами, которые поддерживаются только начиная с определенной версии SDK, а обратную поддержку не завезли(
Подробнее о папках ресурсов можно почитать в [документации от Google](https://developer.android.com/guide/topics/resources/providing-resources)

В данном случае в нашей верстке ничего специфичного для SDK-28 нет, поэтому мы можем смело уделить всю папку `layout-v28` без последствий и поддерживать только `layout`

PavelRomashin replied:
У меня личный телефон, на котором я тестирую то,что написал, стягивает именно этот лейаут, приходится поэтому дублировать)

kirillNay commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -64,6 +100,59 @@ class SearchActivity : AppCompatActivity() {
+                updateButton.visibility = View.VISIBLE
+                errorImage.setImageResource(R.drawable.network_error)
+                errorText.text = text
+                updateButton.setOnClickListener { search() }
```
⏫ Чтобы не выставлять каждый раз при показе ошибки слушателя для кнопки `updateButton`, можно сделать это один раз в методе `onCreate`

kirillNay commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -64,6 +100,59 @@ class SearchActivity : AppCompatActivity() {
+                errorImage.setImageResource(R.drawable.not_found)
+                errorText.text = text
+            } else {
+                updateButton.visibility = View.VISIBLE
```
⏫ Есть очень хороший котлиновский extension у класса `View` -  `isVisible`. С его помощью можно менять видимость элемента лишь изменением значения `Boolean`
`View.isVisible = true` -> соответствует `View.VISIBLE`
`View.isVisible = false` -> соответствует `View.GONE`

kirillNay commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -64,6 +100,59 @@ class SearchActivity : AppCompatActivity() {
+                        if (response.code() == 200) {
+                            trackLibrary.clear()
+                            if (response.body()?.results?.isNotEmpty() == true) {
+                                trackLibrary.addAll(response.body()?.results!!)
```
⚠️  Если мы получили ошибку на какой-то запрос (или ничего не нашли), и после этого решили изменить запрос или нажать на кнопку "**Обновить**", то после успешного ответа список треков не отобразится на экране, т.к. видимость заглушек не меняется

kirillNay commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -60,8 +62,56 @@
+                    android:textFontWeight="400"
+                    android:textSize="19sp" />
+
+                <Button
```
⚠️ Кнопка "**Обновить**" в темной теме имеет белый текст, который не виден на белом фоне. Советую вынести атрибуты кнопки в стили и добавить стиль для темной темы

<img width="330" alt="image" src="https://user-images.githubusercontent.com/56832972/221797385-11a45edf-8b95-45e6-81a2-4a19f0f8ec12.png">

PavelRomashin committed:
Исправление замечаний:
1) Строки вынесены в ресурсы
2) Поправлен метод, кодирующий длительность трека и тип переменной в длительности трека
3)Sealed class и сохранение данных при повороте как реализовать грамотно, пока не понял, а сроки поджимают
4) Переменная trackLibrary перенесена в SearchActivity.kt
5)Добавлена видимость recycler view при поиске. Теперь при обновлении или изменении запроса отображается список
6)исправлено цветовое отображение кнопки обновить
7)видимость элементов теперь идет через булевы числа
